src/platform/silabs/tracing/SilabsTracing.cpp

```diff
@@ -0,0 +1,277 @@
+    tracker.mType      = OperationType::kBegin;
+    tracker.mError     = CHIP_NO_ERROR;
+
+    auto & watermark = mWatermarks[static_cast<size_t>(aOperation)];
+    watermark.mTotalCount++;
+
+    OutputTrace(tracker);
+}
+
+void SilabsTracer::TimeTraceEnd(TimeTraceOperation aOperation, CHIP_ERROR error)
+{
+    auto & tracker   = mLatestTimeTrackers[static_cast<size_t>(aOperation)];
+    tracker.mEndTime = System::SystemClock().GetMonotonicTimestamp();
+    tracker.mType    = OperationType::kEnd;
+    tracker.mError   = error;
+
+    if (error == CHIP_NO_ERROR)
+    {
+        // Calculate the duration and update the time tracker
+        auto duration = tracker.mEndTime - tracker.mStartTime;
+
+        auto & watermark = mWatermarks[static_cast<size_t>(aOperation)];
+        watermark.mSuccessfullCount++;
+        watermark.mMovingAverage = System::Clock::Milliseconds32(
+            (watermark.mMovingAverage.count() * (watermark.mSuccessfullCount - 1) + duration.count()) /
+            watermark.mSuccessfullCount);
+
+        if (duration > watermark.mMaxTimeMs)
+        {
+            watermark.mMaxTimeMs = System::Clock::Milliseconds32(duration);
+        }
+
+        if (watermark.mMinTimeMs.count() == 0 || duration < watermark.mMinTimeMs)
+        {
+            watermark.mMinTimeMs = System::Clock::Milliseconds32(duration);
+        }
+
+        if (duration > watermark.mMovingAverage)
+        {
+            watermark.mCountAboveAvg++;
+        }
+    }
+
+    OutputTrace(tracker);
+}
+
+void SilabsTracer::TimeTraceInstant(TimeTraceOperation aOperation, CHIP_ERROR error)
+{
+    TimeTracker tracker;
+    tracker.mStartTime = System::SystemClock().GetMonotonicTimestamp();
+    tracker.mEndTime   = tracker.mStartTime;
+    tracker.mOperation = aOperation;
+    tracker.mType      = OperationType::kInstant;
+    tracker.mError     = error;
+    OutputTrace(tracker);
+}
+
+CHIP_ERROR SilabsTracer::OutputTimeTracker(const TimeTracker & tracker)
+{
+    VerifyOrReturnError(isLogInitialized(), CHIP_ERROR_UNINITIALIZED);
+    ChipLogProgress(DeviceLayer,
+                    "TimeTracker - StartTime: %" PRIu32 ", EndTime: %" PRIu32 ", Operation: %" PRIu8 ", Type: %" PRIu8
+                    ", Error: %" CHIP_ERROR_FORMAT,
+                    tracker.mStartTime.count(), tracker.mEndTime.count(), static_cast<uint8_t>(tracker.mOperation),
+                    static_cast<uint8_t>(tracker.mType), tracker.mError.Format());
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SilabsTracer::OutputTrace(const TimeTracker & tracker)
+{
+    // We allow error here as we want to buffer even if the logs are currently uninitialized
+    OutputTimeTracker(tracker);
+
+    if (mBufferedTrackerCount < kMaxBufferedTraces - 1)
+    {
+        mTimeTrackerList.Insert(tracker);
+        mBufferedTrackerCount++;
+    }
+    else
+    {
+        // Save a tracker with 0xFFFF to indicate that the buffer is full
+        TimeTracker resourceExhaustedTracker = tracker;
```

Comment:
will we add a failure trace each time after the first one?

Reply:
No, we should only do this once. Will eddit.